desc/optimize/utils.py

```diff
@@ -2,76 +2,97 @@
     -------
     L : ndarray
         Lower triangular cholesky factor, such that L@L.T ~ A
 
     """
+    assert int(maxiter) == maxiter
+    maxiter = int(maxiter)
+    k = 2**maxiter  # number of values to try
     A = jnp.asarray(A)
     n = A.shape[0]
-
-    delta = jnp.finfo(A.dtype).eps * jnp.linalg.norm(A, "fro")
-    xi = jnp.max(jnp.abs(A - jnp.diag(jnp.diag(A))))
-    eta = jnp.max(jnp.abs(jnp.diag(A)))
-    beta = jnp.sqrt(jnp.asarray([eta, xi / n, jnp.finfo(A.dtype).eps]).max())
-
-    L = jnp.zeros_like(A)
-    D = jnp.zeros(n)
-    c = jnp.zeros_like(A)
-
-    def sum_slice(D, L, i, j, kmax):
-        s = 0
-        bodyfun = lambda k, s: s + D[k] * L[i, k] * L[j, k]
-        return fori_loop(0, kmax, bodyfun, s)
-
-    def inner_loop(j, cLDi):
-        c, L, D, i = cLDi
-        s = sum_slice(D, L, i, j, j)
-        c = put(c, (i, j), A[i, j] - s)
-
-        def truefun(args):
-            i, c, D = args
-            theta = jnp.max(jnp.abs(c[i, :]))
-            D = put(D, i, jnp.asarray([abs(c[i, i]), (theta / beta) ** 2, delta]).max())
-            return i, c, D
-
-        def falsefun(args):
-            return args
-
-        i, c, D = cond(i == j, truefun, falsefun, (i, c, D))
-        L = put(L, (i, j), c[i, j] / D[j])
-
-        return c, L, D, i
-
-    def outer_loop(i, cLD):
-        c, L, D = cLD
-        c, L, D, i = fori_loop(0, i + 1, inner_loop, (c, L, D, i))
-        return c, L, D
-
-    c, L, D = fori_loop(0, n, outer_loop, (c, L, D))
-    return L * jnp.sqrt(D)
+    eye = jnp.eye(n)
+    # upper and lower bounds on eig(A)
+    lb, ub = gershgorin_bounds(A)
+    # upper bound on log(alpha) such that A + alpha*I > 0, ie we know alpha < ub
+    # lower bound on eig(A) = upper bound on alpha, +1 in log scale to make sure
+    # that it's actually greater than the maximum alpha
+    ub = jnp.log10(jnp.abs(lb)) + 1
+    # we know alpha > 0 because otherwise initial factorization would have succeeded
+    # but we'd like to be a bit better (in log scaling). This is just a heuristic but
+    # seems ok in practice
+    m = jnp.mean(jnp.abs(A))
+    lb = ub - 2 * abs(ub) - abs(jnp.log10(m))
+    # values to try
+    alphas = jnp.logspace(lb, ub, k)
+    kbest = k // 2
+    klow = 0
+    khigh = k
+    # first we try alpha = max, which we know will succeed by gershgorin bounds
+    # but might be too big a correction, so then we try to reduce it while keeping
+    # A + alpha*I positive definite
+    Lbest = jnp.linalg.cholesky(A + alphas[k] * eye)
+    for i in range(maxiter):
+        L = jnp.linalg.cholesky(A + alphas[kbest] * eye)
+        # check if it succeeded
+        isnan = jnp.any(jnp.isnan(L))
+        # adjust bounds for correction
+        klow = isnan * kbest + (1 - isnan) * klow
+        khigh = isnan * khigh + (1 - isnan) * kbest
+        kbest = (klow + khigh) // 2
+        # if it succeeded, mark it as the best so far
+        Lbest = cond(isnan, lambda _: Lbest, lambda _: L, 1)
+    return Lbest
```

Comment:
So this starts with the most conservative correction, tries all possible values for alpha (maxiter different values), and then returns the smallest correction that worked? 
Wouldn't it be better to start with the smallest value of alpha, and then return the first one that works? 
That way we won't always have to do `maxiter` iterations -- we could get lucky on the first or second try.

Reply:
On that same train of thought, this doesn't actually need to be a separate function from `chol(A)` if you start with trying alpha=0.

Reply:
In that case in the worst case you need 1 extra attempted factorization, and in testing that seemed to be fairly common. If you start with the smallest one and it doesn't work, the next one you'd try would be the middle one, and if that works you'd try a quarter etc, so in almost all cases it will still take the full maxiter. Starting with the smallest one in the best case terminates after 1 iteration, in the worst case requires maxiter+1, and it seemed like the worst case was actually fairly common (because we're only searching over a log scale).